### PR TITLE
Email attachment

### PR DESCRIPTION
**Pull request by saurabhketech**

### CODE REVIEW

File: src/controllers/fetch_email.js [hunk 1]
```diff
@@ -1,7 +1,31 @@
 import BaseAPIController from "./BaseAPIController";
 import MailProvider from "../providers/MailProvider";
 import * as _ from "lodash";
-
+import Imap from "imap";
```
Comment (mageexxcel) [inline src/controllers/fetch_email.js]: this filename was changed in a different pR

File: src/controllers/fetch_email.js [hunk 2]
```diff
@@ -315,10 +336,164 @@ export class FetchController extends BaseAPIController {
...
+                        });
+                } else {
+                    res.json({ status: 0, message: 'mongo_id not found in database' });
+                }
+            }
+        });
+    }
+}
+
+function get_attachment(imap, uid) {
```
Comment (mageexxcel) [inline src/controllers/fetch_email.js]: remove from controller

File: src/controllers/fetch_email.js [hunk 3]
```diff
@@ -315,10 +336,164 @@ export class FetchController extends BaseAPIController {
...
+                    reject("Fetch error: " + err);
+                });
+                f.once("end", () => {
+                    console.log("Done fetching all messages!");
+                    imap.end();
+                });
+            });
+        });
+
+        function findAttachmentParts(struct, attachments) {
```
Comment (mageexxcel) [inline src/controllers/fetch_email.js]: dont add function inside another function

File: src/controllers/fetch_email.js [hunk 4]
```diff
@@ -1,7 +1,31 @@
...
-
+import Imap from "imap";
+import in_array from "in_array";
+import fs from "fs";
+import base64 from "base64-stream";
+import path from "path";
+import multer from "multer";
+import google from "googleapis";
+var OAuth2 = google.auth.OAuth2;
+var upload = multer({
```
Comment (mageexxcel) [inline src/controllers/fetch_email.js]: where is multer used?

File: src/controllers/fetch_email.js [hunk 5]
```diff
@@ -315,10 +336,164 @@ export class FetchController extends BaseAPIController {
...
+                    attachments.push(struct[i]);
+                }
+            }
+            return attachments;
+        }
+
+        function buildAttMessageFunction(attachment, uid, flag) {
+            var filename = attachment.disposition.params.filename;
+            var encoding = attachment.encoding;
+            var filepath = path.join(__dirname, "/./uploads/", filename);
```
Comment (mageexxcel) [inline src/controllers/fetch_email.js]: why is this required /./

File: src/controllers/fetch_email.js [hunk 6]
```diff
@@ -315,10 +336,164 @@ export class FetchController extends BaseAPIController {
...
+                        stream.pipe(base64.decode()).pipe(writeStream);
+                    } else {
+                        stream.pipe(writeStream);
+                    }
+                    fs.readFile(filepath, {
+                        encoding: "utf8"
+                    }, (error, data) => {
+                        var fileMetadata = {
+                            title: filename,
+                            mimeType: "text/plain/javascript/html/csv/application/pdf"
```
Comment (mageexxcel) [inline src/controllers/fetch_email.js]: why is this always hard coded to pdf?

File: src/controllers/fetch_email.js [hunk 7]
```diff
@@ -315,10 +336,164 @@ export class FetchController extends BaseAPIController {
...
+                });
+                msg.once("end", () => {
+                    console.log(prefix + "Finished attachment %s", filename);
+                });
+            };
+        }
+        imap.once("error", (err) => {
+            reject(err);
+        });
+        imap.once("end", () => {
```
Comment (mageexxcel) [inline src/controllers/fetch_email.js]: does the imap connection close automatically? i don't see any code to close connection

File: src/inbox.js [hunk 8]
```diff
@@ -36,7 +36,7 @@ module.exports = {
                             var yesterday = new Date();
                             yesterday.setTime(Date.now() - delay);
                             yesterday = yesterday.toISOString();
-                            imap.search(["ALL", ["BEFORE", yesterday]], function(err, results) {
+                            imap.search(["ALL", ["SINCE", yesterday]], function(err, results) {
```
Comment (mageexxcel) [inline src/inbox.js]: why is this changed?

File: src/models/smtp.js [hunk 9]
```diff
@@ -37,6 +34,23 @@ export default function(sequelize, DataTypes) {
...
 
             }
+        },
+        classMethods: {
+            smtpTest(email) {
+                return new Promise((resolve, reject) => {
+                    this.findOne({ where: { email: email } })
+                        .then((data) => {
+                            if (data) {
+                                sequelize.query(`UPDATE SMTP  SET status = (case when email = "${email}" then 1 else 0 end);`)
```
Comment (mageexxcel) [inline src/models/smtp.js]: this was fixed in a different PR

File: src/controllers/fetch_email.js [hunk 10]
```diff
@@ -315,10 +336,164 @@ export class FetchController extends BaseAPIController {
...
+            if (error) {
+                next(new Error(error));
+            } else {
+                if (data) {
+                    let sender_mail = data.get("sender_mail");
+                    let to = data.get("to");
+                    let uid = data.get("uid");
+                    this._db.Imap.findOne({ email: to })
+                        .then((data) => {
+                            var imap = new Imap({
```
Comment (mageexxcel) [inline src/controllers/fetch_email.js]: this code has been used in many places

i.e find active imap setting and start connection with image. 

i have seen same code copy pasted 3 times already.

make a single service for all imap codes

File: src/service/cron.js [hunk 11]
```diff
@@ -3,10 +3,10 @@ var CronJob = require("cron").CronJob;
 
 export class CronController {
     cron(email) {
-        // new CronJob("*/15 * * * *", function() {
-            inbox.fetch_email(email); // running this function every 15 min
-            // inbox.Before_date_email(email, new Date());
-        // }, null, true);
+        new CronJob("*/15 * * * *", function() {
+            inbox.fetchEmail(email); // running this function every 15 min
+            inbox.beforeDateEmail(email, new Date());
```
Comment (mageexxcel) [inline src/service/cron.js]: this is wrong. here you are passing new Date() everytime.

how will this function work. 

this will always fetch the same again and again. we need to go back in dates, not ahead

File: src/controllers/fetch_email.js (hunk 1 above)
Comment (saurabhketech) [reply src/controllers/fetch_email.js]: filename changed to fetchEmail.js

File: src/controllers/fetch_email.js (hunk 4 above)
Comment (saurabhketech) [reply src/controllers/fetch_email.js]: done

File: src/controllers/fetch_email.js (hunk 10 above)
Comment (saurabhketech) [reply src/controllers/fetch_email.js]: done

File: src/controllers/fetch_email.js (hunk 2 above)
Comment (saurabhketech) [reply src/controllers/fetch_email.js]: done

File: src/controllers/fetch_email.js (hunk 3 above)
Comment (saurabhketech) [reply src/controllers/fetch_email.js]: done

File: src/controllers/fetch_email.js (hunk 7 above)
Comment (saurabhketech) [reply src/controllers/fetch_email.js]: imap close  connection when imap.end() is call

File: src/inbox.js (hunk 8 above)
Comment (saurabhketech) [reply src/inbox.js]: we are fetching data from since yesterday, all the emails  from yesterday

File: src/models/smtp.js (hunk 9 above)
Comment (saurabhketech) [reply src/models/smtp.js]: done

File: src/service/cron.js (hunk 11 above)
Comment (saurabhketech) [reply src/service/cron.js]: we are creating two function 1)fetchEmail fetch data from yesterday 
2)beforeDateEmail fetch data from before date

File: src/controllers/fetch_email.js (hunk 5 above)
Comment (saurabhketech) [reply src/controllers/fetch_email.js]: done

File: src/controllers/fetch_email.js (hunk 6 above)
Comment (saurabhketech) [reply src/controllers/fetch_email.js]: done
